src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -265,107 +295,122 @@ private static void CopySignatureField(byte[] signatureField, byte[] response, i
             }
         }
 
-        private void FreeKey()
+        public override int KeySize
         {
-            if (_key != null && _key.IsValueCreated)
+            get
             {
-                SafeEcKeyHandle handle = _key.Value;
-
-                if (handle != null)
+                if (_curve != null)
                 {
-                    handle.Dispose();
+                    // Load the key so we don't return 0
+                    SafeEcKeyHandle key = _key.Value;
+                    Debug.Assert(_curve == null);
                 }
+
+                return base.KeySize;
+            }
+            set
+            {
+                if (KeySize == value)
+                    return;
+
+                // Set the KeySize first so that an invalid value doesn't throw away the key
+                base.KeySize = value;
+
+                FreeKey();
+                _key = new Lazy<SafeEcKeyHandle>(GenerateKeyLazy);
             }
         }
 
-        private static int GetKeySize(SafeEcKeyHandle ecKeyHandle)
+        public override void GenerateKey(ECCurve curve)
         {
-            int nid = Interop.Crypto.EcKeyGetCurveName(ecKeyHandle);
-            int keySize = 0;
+            curve.Validate();
+            FreeKey();
 
-            for (int i = 0; i < s_supportedAlgorithms.Length; i++)
+            if (curve.IsNamed)
             {
-                if (s_supportedAlgorithms[i].Nid == nid)
-                {
-                    keySize = s_supportedAlgorithms[i].KeySize;
-                    break;
-                }
+                _curve = curve;
+                _key = new Lazy<SafeEcKeyHandle>(GenerateKeyLazy);
+                // Generate the key
+                var dummy = _key.Value;
             }
-
-            if (keySize == 0)
+            else if (curve.IsExplicit)
             {
-                string curveNameOid = Interop.Crypto.GetOidValue(Interop.Crypto.ObjNid2Obj(nid));
-                throw new NotSupportedException(SR.Format(SR.Cryptography_UnsupportedEcKeyAlgorithm, curveNameOid));
+                SafeEcKeyHandle key = Interop.Crypto.EcKeyCreateByExplicitCurve(curve);
+                if (!Interop.Crypto.EcKeyGenerateKey(key))
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+                SetKey(key);
+            }
+            else
+            {
+                throw new NotSupportedException(string.Format(SR.Cryptography_InvalidCurve, curve.CurveType.ToString()));
             }
-
-            return keySize;
         }
 
-        private SafeEcKeyHandle GenerateKey()
+        private SafeEcKeyHandle GenerateKeyLazy()
         {
-            int keySize = KeySize;
-            for (int i = 0; i < s_supportedAlgorithms.Length; i++)
+            string oid = null;
+            if (_curve != null)
             {
-                if (keySize == s_supportedAlgorithms[i].KeySize)
+                Debug.Assert(_curve.Value.IsNamed);
```

Comment:
At that time, only named curves were lazy-loaded. Changing per other comments to not do lazy ever